andela/eagle-bn-backend
# #168781812 trips request

#### What does this PR do?
- User can make a one-way trip request
- User can make a return-trip request
- user can make a multi-city trip request
- connecting roles with the users
#### Description of Task to be completed?
when user wants to make a one-way trip: returnTime is not required
`req.body:`
 ```
{
 country: [countryOfOrigin],
 city: [cityOfOrigin],
 timeZone: [timeZone of the user],
 trips: [
     country: [countryOfDestination],
     city: [cityOfDestination],
     departureTime: [format('M/D/YYYY h:mm a, e.g: 10/25/2019 10:55 pm')],
     reason: [purposeOfTheTrip]
]
}
```
when user wants to make a return trip request: returnTime is required
`req.body:` 
```
{
 country: [countryOfOrigin],
 city: [cityOfOrigin],
 timeZone: [timeZone of the user],
 returnTime: [format('M/D/YYYY h:mm a, e.g: 10/25/2019 10:55 pm')],
 trips: [
     country: [countryOfDestination],
     city: [cityOfDestination],
     departureTime: [format('M/D/YYYY h:mm a, e.g: 10/25/2019 10:55 pm')],
     reason: [purposeOfTheTrip]
]
}
```
when user wants to make a multi-city trip request: returnTime is optional
`req.body:` 
```
{
 country: [countryOfOrigin],
 city: [cityOfOrigin],
 timeZone: [timeZone of the user],
 returnTime: optional || [format('M/D/YYYY h:mm a, e.g: 10/25/2019 10:55 pm')],
 trips: [
     country: [countryOfDestination],
     city: [cityOfDestination],
     departureTime: [format('M/D/YYYY h:mm a, e.g: 10/25/2019 10:55 pm')],
     reason: [purposeOfTheTrip],
   ............. // he can make as many as he want
]
}
```

`Assertion on trips`:
```
- user should make an invalid time span(from departureTime to returnTime);
- user should not make trips in the same destination
- user should provide a valid time zone,(we have a set of valid time zone)
- country property should be shortand, e.g: US instead of united state or UG instead of Uganda
```
#### How should this be manually tested?
*clone the repo, install Postgresql, install packages and setup your .env by following .env.sample*
```
- routes for making trip request= post:{hostAddress}:{hostPort}/api/v1/requests
- follow the description above on how you should format the body in postman
there is a predefined test under `src/tests/requests.post.spec.js`
you can run test with `npm test`
```
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[story](https://www.pivotaltracker.com/story/show/168781812)
#### Screenshots (if appropriate)
*to be provided.....*
#### Questions:
N/A
